### ISSUE TITLE
FreeBSD

### ISSUE THREAD

**Issue by ZWS2014** [scaleway/image-proposals#11]
## Vendor URLs

* https://www.freebsd.org/
* https://github.com/freebsd/freebsd
* https://www.freebsd.org/docs.html

## Use cases / something else ?

Jails, ZFS, security (auditing, timely patching, system robustness), performance, well managed system & port updates, the list goes on ...

**Comment by moul** [comment]
* FreeBSD on x86_64 (C2S, C2M, C2L, VC1S, VC1M, VC1L) needs an alternative way to store data or an NBD driver for FreeBSD, and more work on the boot system
* FreeBSD on armhf (C1) requires a valid kernel (with full support of armada xp processors, and network), a valid init ramdisk, an alternative way to store data or an NBD driver for FreeBSD, and some work on the boot system too

we can expect the x86_64 version to be released before the armhf one which seems complicated and depending a lot on the community

**Comment by ZWS2014** [comment]
@moul Thanks for the update. Would you have any idea of a possible ETA for the x86_64 version, which is what I'd be looking for?

I had some success running FreeBSD 10.3 in KVM/QEUM instance with attached volumes as ZFS storage (though not primarily for a data integrity use case, for obvious reasons), and I also managed to install FreeBSD on a volume attached to an Ubuntu installation through the Scaleway console. However, there were a number of issues using these methods and neither produced a production ready solution.

Thank you to Scaleway for planning to support FreeBSD - that would be fantastic. I'd imagine it'd be a good selling point, as the BSDs still seem under provided in the VPS space (DigitalOcean being a notable exception). The power, scalability and price point that Scaleway offers in both VPS/Dedicated might be a good fit for FreeBSD installs; whilst the number of FreeBSD users may be relatively small when compared to Linux, the user base is growing and I'd suggest the time could be ripe to develop a reputation as a "go-to" solution for FreeBSD hosting, provided it's done right.

**Comment by myfreeweb** [comment]
> an alternative way to store data or an NBD driver for FreeBSD

Wait, do you attach all SSD storage over the network? You don't do local storage? :(

I really like your pricing (€2.99 for 2GB RAM! currently paying $5 for 512MB, I'd really like more RAM) but moving from local SSD to networked doesn't sound very good.

> on armhf (C1) requires a valid kernel (with full support of armada xp processors, and network)

There's [a kernel config](https://github.com/freebsd/freebsd/blob/master/sys/arm/conf/ARMADAXP) and [initialization code or something](https://github.com/freebsd/freebsd/blob/master/sys/arm/mv/armadaxp/armadaxp.c) for Armada XP. Marvell Gigabit Ethernet is supported as well. Not sure how tested it is. I don't have anything with an Armada, I only have some Raspberry Pis :D

> a valid init ramdisk

I don't think FreeBSD uses init ramdisks anywhere at all.

**Comment by justinclift** [comment]
> I really like your pricing (€2.99 for 2GB RAM! currently paying $5 for 512MB, I'd really like more RAM) but moving from local SSD to networked doesn't sound very good.

Depending on the storage network, it can still be a pretty quick solution.  I've personally run a lot of my development lab environments over the last few years with storage on remote boxes.  As long as the network is 10GbE or better, and not totally overloaded :wink:, it's not possible for an end user to tell it's remote.

If you have some seriously I/O intensive apps running against it though, then yeah that might need measuring and factoring in. :smile:

**Comment by myfreeweb** [comment]
[Looks like you have AArch64 servers now, based on ThunderX](https://blog.online.net/2017/04/27/scaleway-disruptive-armv8-cloud-servers/)… This is nice. And ThunderX is literally the reference platform for FreeBSD/arm64! I hope you'll be able to support FreeBSD on that…

**Comment by jnordberg** [comment]
Can't wait for some bare metal FreeBSD love! 😈 👍

**Comment by asomers** [comment]
I don't suppose there's any possibility that FreeBSD could mount its storage with iSCSI instead of NBD, is there?  FreeBSD already has an in-kernel iSCSI client.

**Comment by myfreeweb** [comment]
@jnordberg probably won't be bare metal since the ThunderX's are 48 core monsters, that calls for VPSes :)

**Comment by justinclift** [comment]
@myfreeweb iSCSI is a network transport.  Guessing you misread as "SCSI"? :wink:

**Comment by asomers** [comment]
@justinclift are you talking to me?  I meant iSCSI when I said it.  FreeBSD has an in-kernel iSCSI initiator.  Ffunctionally it's more or less equivalent to NBD.

**Comment by myfreeweb** [comment]
@justinclift I'm not talking about iSCSI, I'm replying to "Can't wait for some bare metal FreeBSD love" :D

**Comment by jnordberg** [comment]
I'd be happy to see any FreeBSD image, even if it's just a VPS one 😄

**Comment by justinclift** [comment]
@myfreeweb Oops, sorry.  Thought you were replying to @asomers.  Wasn't paying close enough attention. :wink:

**Comment by vadzimt** [comment]
No FreeBSD love :heart: yet?

**Comment by vadmeste** [comment]
:+1:

**Comment by nerzhul** [comment]
I hope baremetal freebsd can happen soon to replace my dedibox 2015 :)

**Comment by tehfink** [comment]
FreeBSD FTW!

**Comment by fredt34** [comment]
Looks like it was done here: https://community.online.net/t/freebsd-on-a-vc1/2385

**Comment by myfreeweb** [comment]
So looking at the link above, the x86_64 VMs do boot into UEFI… What about aarch64?

**Comment by justinclift** [comment]
@myfreeweb With this:

> One remaining issue: incoming TCP connections (like ssh, or simply testing with nc) from the public IPv4 address don’t work — I see incoming packets in tcpdump but no replies o_0 (IPv6 works perfectly, at least)

What's that from the point of view of?  The IP address on the interfaces themselves generally seem to be something internal to Scaleway, but somewhere in the Scaleway network layer they appear to be translated (NAT of some variety) to the public version.  Is there any chance you're expecting the public IP address to be used in packets hitting/sending from your instance interface?

**Comment by myfreeweb** [comment]
Yeah, I don't know *what* could be expecting the IP to be the public one — in the FreeBSD x86 thread, no one mentioned any issues with that…

**Comment by myfreeweb** [comment]
Okay, turns out the broadcast address should be `255.255.255.255`, now works fine! :)

**Comment by justinclift** [comment]
Awesome. :smile:

**Comment by marco-m** [comment]
@moul any news about making an official image available? Merci!
